packages/engine/Source/DataSources/FeatureLayer.js

```diff
@@ -0,0 +1,213 @@
+import Rectangle from "../Core/Rectangle.js";
+import RuntimeError from "../Core/RuntimeError.js";
+import defined from "../Core/defined.js";
+import PolygonGeometry from "../Core/PolygonGeometry.js";
+import Feature from "./Feature.js";
+import AssociativeArray from "../Core/AssociativeArray.js";
+import CustomDataSource from "./CustomDataSource.js";
+import parseGeoJson from "./parseGeoJson.js";
+import Ellipsoid from "../Core/Ellipsoid.js";
+import Cartesian3 from "../Core/Cartesian3.js";
+
+/**
+ * @import FeatureProvider from "./FeatureProvider";
+ * @import FrameState from "../Scene/FrameState";
+ * @import OGCFeaturesProvider from "./OgcFeatureProvider.js";
+ */
+
+/**
+ * A class for things
+ * @constructor
+ * @param {FeatureProvider | OGCFeaturesProvider} provider
+ * @param {Object} options
+ */
+function FeatureLayer(provider, options) {
+  this.provider = provider;
+
+  this.id = options.id;
+  // this.name = options.name;
+  this._show = options.show ?? true;
+
+  this.rectangle = options.rectangle ?? new Rectangle();
+
+  // TODO: convert this to a PrimitiveCollection
+  // this._dataSource = new CustomDataSource();
+  // // this._primitives = new EntityCollection();
+  // /** @type {EntityCollection} */
+  // this._primitives = this._dataSource.entities;
+  // this._dataSource = new GeoJsonDataSource(options.id);
+  this._customDataSource = new CustomDataSource();
+  // this._primitives = this._dataSource.entities;
+  this._primitives = [];
+
+  this.features = options.features ?? new AssociativeArray();
```

Comment:
I couldn't find an issue for it, but `AssociativeArray` can probably be replaced with native `Map` at this point. It's easy an efficient to pull the values as an array through built-in methods.

Reply:
Also, and perhaps more importantly, we should be smart about how we store data for lots and lots of features. There's potential for a lot of duplication across the graphical representation (whether than be Entities or Primitives) and a raw array. We likely want to release the raw features once the representations have been created.